dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright 2021. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{collections::HashMap, fmt::Debug, hash::Hash, time::Duration};
+
+use log::*;
+use tari_shutdown::ShutdownSignal;
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    task::JoinHandle,
+};
+
+use super::hotstuff_error::HotStuffError;
+
+const LOG_TARGET: &str = "tari::dan_layer::pacemaker_worker";
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PacemakerWaitStatus {
+    WaitTimeOut,
+}
+
+async fn send_timeout_message<T: Debug + Send>(
+    wait_over: T,
+    tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+) -> Result<(), HotStuffError> {
+    info!(target: LOG_TARGET, "Sending timeout message for value: {:?}", wait_over);
+    tx_waiter_status
+        .send((wait_over, PacemakerWaitStatus::WaitTimeOut))
+        .await
+        .map_err(|_| HotStuffError::SendError)?;
+    Ok(())
+}
+
+/// A pacemaker service that is responsible for:
+///   1. Receiving [`start_wait`] messages parametrized over a type `T` and start wait up to a max_timeout value.
+///   2. Receiving [`stop_wait`] messages for an already existing [`start_wait`].
+///   3. If it receives a stop message then it should stop the waiting and do nothing else;
+///   4. Otherwise, it should notify the sender of [`start_wait`] message that the max_timeout duration has passed.
+#[derive(Debug)]
+pub struct Pacemaker<T> {
+    rx_start_signal: Receiver<T>,
```

Comment:
Sorry for nitpicking, but can you add a comment to each field what's the purpose. To me and you it's obvious, but for future generations :-)